src/main/scala/com/typesafe/sbt/packager/graalvmnativeimage/GraalVMNativeImagePlugin.scala

```diff
@@ -0,0 +1,191 @@
+            resourceMappings,
+            image,
+            streams
+          )
+      }
+    }
+  )
+
+  private def buildLocal(targetDirectory: File,
+                         binaryName: String,
+                         className: String,
+                         classpathJars: Seq[File],
+                         extraOptions: Seq[String],
+                         log: ProcessLogger): File = {
+    targetDirectory.mkdirs()
+    val command = {
+      val nativeImageArguments = {
+        val classpath = classpathJars.mkString(":")
+        Seq("--class-path", classpath, s"-H:Name=$binaryName") ++ extraOptions ++ Seq(className)
+      }
+      Seq(NativeImageCommand) ++ nativeImageArguments
+    }
+    sys.process.Process(command, targetDirectory) ! log match {
+      case 0 => targetDirectory / binaryName
+      case x => sys.error(s"Failed to run $command, exit status: " + x)
+    }
+  }
+
+  private def buildInDockerContainer(targetDirectory: File,
+                                     binaryName: String,
+                                     className: String,
+                                     classpathJars: Seq[(File, String)],
+                                     extraOptions: Seq[String],
+                                     dockerCommand: Seq[String],
+                                     resources: Seq[(File, String)],
+                                     image: String,
+                                     streams: TaskStreams): File = {
+
+    stage(targetDirectory, classpathJars, resources, streams)
+
+    val command = dockerCommand ++ Seq(
+      "run",
+      "--rm",
+      "-v",
+      s"${targetDirectory.getAbsolutePath}:/opt/graalvm",
+      image,
+      "-cp",
+      classpathJars.map(jar => "/opt/graalvm/stage/" + jar._2).mkString(":"),
+      s"-H:Name=$binaryName"
+    ) ++ extraOptions ++ Seq(className)
+
+    sys.process.Process(command) ! streams.log match {
+      case 0 => targetDirectory / binaryName
+      case x => sys.error(s"Failed to run $command, exit status: " + x)
+    }
+  }
+
+  /**
+    * This can be used to build a custom build image starting from a custom base image. Can be used like so:
+    *
+    * ```
+    * (containerBuildImage in GraalVMNativeImage) := generateContainerBuildImage("my-docker-hub-username/my-graalvm").value
+    * ```
+    *
+    * The passed in docker image must have GraalVM installed and on the PATH, including the gu utility.
+    */
+  def generateContainerBuildImage(baseImage: String): Def.Initialize[Task[Option[String]]] = Def.task {
+    val dockerCommand = (DockerPlugin.autoImport.dockerExecCommand in GraalVMNativeImage).value
+    val streams = Keys.streams.value
+    val target = (Keys.target in GraalVMNativeImage).value
+
+    val (baseName, tag) = baseImage.split(":", 2) match {
+      case Array(n, t) => (n, t)
+      case Array(n)    => (n, "latest")
+    }
+
+    val imageName = s"${baseName.replace('/', '-')}-native-image:$tag"
```

Comment:
Should we rely on tags here? This can be problematic, e.g. if we have two builds running in parallel with different versions of this plugin.

`docker build -q -f <dockerFile> .` builds an untagged image, and prints its ID. We can use that instead.

Reply:
The problem is, I don't want to have to rebuild the image every time I build my graal native image, nor do I want to build separate images for each project. This only rebuilds the image if the tag is not present. This is mentioned in the docs.

I don't think two builds running in parallel (or rather, one build using another builds cached image, which is the real problem) is going to cause a problem. The image produced just needs to execute `native-image` as its entrypoint. It doesn't matter what dockerfile was used to produce that, so it doesn't matter which one gets used. And, if in some rare case it does matter, then it can be overridden.

Besides, we are 100% in control of the docker file, if some future version of native packager changes it in some way that is not compatible with old images generated, then all we have to do is change the generated tag name used by the new version of native packager, and we guarantee no conflicts.

Reply:
> The problem is, I don't want to have to rebuild the image every time I build my graal native image, nor do I want to build separate images for each project. This only rebuilds the image if the tag is not present. This is mentioned in the docs.

That shouldn't be much of an issue in practice, since our `Dockerfile` doesn't use `ADD` and `COPY`, so it should just resolve the layers from cache.

> It doesn't matter what dockerfile was used to produce that, so it doesn't matter which one gets used.

If we later want to add any other feature to this image, this will no longer hold. There's also the fact that the tagged source image can change (especially if the tag is `latest`). Ultimately this is future-proofing on my side, paired with paranoia about all mutable global state. :smile:

> Besides, we are 100% in control of the docker file, if some future version of native packager changes it in some way that is not compatible with old images generated, then all we have to do is change the generated tag name used by the new version of native packager, and we guarantee no conflicts.

It's a fair point that we control the docker file here, so YAGNI applies. However, changing the tag in case of a conflict seems fragile. If someone uses our known tag for some purpose (e.g. to clean up autogenerated images), we are going to have a problem in that scenario. But using the existing tag should work *for now*, so I'm resolving this.